test/Quake/memtoreg-7.qke

```diff
@@ -0,0 +1,102 @@
+// ========================================================================== //
+// Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                 //
+// All rights reserved.                                                       //
+//                                                                            //
+// This source code and the accompanying materials are made available under   //
+// the terms of the Apache License 2.0 which accompanies this distribution.   //
+// ========================================================================== //
+
+// RUN: cudaq-opt --memtoreg %s | FileCheck %s
+
+func.func @__nvqpp__mlirgen__test() attributes {"cudaq-entrypoint", qubitMeasurementFeedback = true} {
+  %false = arith.constant false
+  %c1_i64 = arith.constant 1 : i64
+  %c0_i64 = arith.constant 0 : i64
+  %true = arith.constant true
+  %c2_i64 = arith.constant 2 : i64
+  %0 = quake.alloca !quake.veq<2>
+  %1 = cc.alloca i1
+  cc.store %true, %1 : !cc.ptr<i1>
+  %2 = cc.loop while ((%arg0 = %c0_i64) -> (i64)) {
+    %5 = arith.cmpi slt, %arg0, %c2_i64 : i64
+    cc.condition %5(%arg0 : i64)
+  } do {
+  ^bb0(%arg0: i64):
+    %5 = quake.extract_ref %0[%arg0] : (!quake.veq<2>, i64) -> !quake.ref
+    %measOut = quake.mz %5 name "res" : (!quake.ref) -> !quake.measure
+    %6 = quake.discriminate %measOut : (!quake.measure) -> i1
+    cc.store %6, %1 : !cc.ptr<i1>
+    %7 = cc.load %1 : !cc.ptr<i1>
+    %8 = arith.cmpi eq, %7, %false : i1
+    cc.if(%8) {
+      %measOut_0 = quake.mz %0 name "inner_mz" : (!quake.veq<2>) -> !cc.stdvec<!quake.measure>
+      %9 = quake.discriminate %measOut_0 : (!cc.stdvec<!quake.measure>) -> !cc.stdvec<i1>
+      cc.scope {
+        %10 = cc.alloca !cc.stdvec<i1>
+        cc.store %9, %10 : !cc.ptr<!cc.stdvec<i1>>
+      }
+    }
+    cc.continue %arg0 : i64
+  } step {
+  ^bb0(%arg0: i64):
+    %5 = arith.addi %arg0, %c1_i64 : i64
+    cc.continue %5 : i64
+  } {invariant}
+  %3 = cc.load %1 : !cc.ptr<i1>
+  %4 = arith.cmpi eq, %3, %true : i1
+  cc.if(%4) {
+    %measOut = quake.mz %0 name "outer_mz" : (!quake.veq<2>) -> !cc.stdvec<!quake.measure>
+    %5 = quake.discriminate %measOut : (!cc.stdvec<!quake.measure>) -> !cc.stdvec<i1>
+    cc.scope {
+      %6 = cc.alloca !cc.stdvec<i1>
+      cc.store %5, %6 : !cc.ptr<!cc.stdvec<i1>>
+    }
+  }
+  return
+}
+
+// CHECK-LABEL:   func.func @__nvqpp__mlirgen__test() attributes {"cudaq-entrypoint", qubitMeasurementFeedback = true} {
+// CHECK-DAG:       %[[VAL_0:.*]] = arith.constant false
+// CHECK-DAG:       %[[VAL_1:.*]] = arith.constant 1 : i64
+// CHECK-DAG:       %[[VAL_2:.*]] = arith.constant 0 : i64
+// CHECK-DAG:       %[[VAL_3:.*]] = arith.constant true
+// CHECK-DAG:       %[[VAL_4:.*]] = arith.constant 2 : i64
+// CHECK-DAG:       %[[VAL_5:.*]] = quake.alloca !quake.veq<2>
+// CHECK-DAG:       %[[VAL_6:.*]] = cc.undef i1
```

Comment:
This test doesn't run on `main` so it's a difficult for me to gauge what the new changes are... of course it passes now.
For example, I don't understand how / why this is `undef`.

Reply:
This is mem2reg. We turn references into values. Like a memory location is undefined until we store something to it, a value chain starts out as undefined until we compute something.